moquette-0.16/broker/src/main/java/io/moquette/broker/subscriptions/CTrie.java

```diff
@@ -0,0 +1,233 @@
+            subscriptions.addAll(cnode.subscriptions);
+        }
+        for (INode subInode : cnode.allChildren()) {
+            subscriptions.addAll(recursiveMatch(remainingTopic, subInode));
+        }
+        return subscriptions;
+    }
+
+    public void addToTree(Subscription newSubscription) {
+        Action res;
+        do {
+            res = insert(newSubscription.topicFilter, this.root, newSubscription);
+        } while (res == Action.REPEAT);
+    }
+
+    private Action insert(Topic topic, final INode inode, Subscription newSubscription) {
+        Token token = topic.headToken();
+        if (!topic.isEmpty() && inode.mainNode().anyChildrenMatch(token)) {
+            Topic remainingTopic = topic.exceptHeadToken();
+            INode nextInode = inode.mainNode().childOf(token);
+            return insert(remainingTopic, nextInode, newSubscription);
+        } else {
+            if (topic.isEmpty()) {
+                return insertSubscription(inode, newSubscription);
+            } else {
+                return createNodeAndInsertSubscription(topic, inode, newSubscription);
+            }
+        }
+    }
+
+    private Action insertSubscription(INode inode, Subscription newSubscription) {
+        CNode cnode = inode.mainNode();
+        CNode updatedCnode = cnode.copy().addSubscription(newSubscription);
+        if (inode.compareAndSet(cnode, updatedCnode)) {
+            return Action.OK;
+        } else {
+            return Action.REPEAT;
+        }
+    }
+
+    private Action createNodeAndInsertSubscription(Topic topic, INode inode, Subscription newSubscription) {
+        INode newInode = createPathRec(topic, newSubscription);
+        CNode cnode = inode.mainNode();
+        CNode updatedCnode = cnode.copy();
+        updatedCnode.add(newInode);
+
+        return inode.compareAndSet(cnode, updatedCnode) ? Action.OK : Action.REPEAT;
+    }
+
+    private INode createPathRec(Topic topic, Subscription newSubscription) {
+        Topic remainingTopic = topic.exceptHeadToken();
+        if (!remainingTopic.isEmpty()) {
+            INode inode = createPathRec(remainingTopic, newSubscription);
+            CNode cnode = new CNode();
+            cnode.setToken(topic.headToken());
+            cnode.add(inode);
+            return new INode(cnode);
+        } else {
+            return createLeafNodes(topic.headToken(), newSubscription);
+        }
+    }
+
+    private INode createLeafNodes(Token token, Subscription newSubscription) {
+        CNode newLeafCnode = new CNode();
+        newLeafCnode.setToken(token);
+        newLeafCnode.addSubscription(newSubscription);
+
+        return new INode(newLeafCnode);
+    }
+
+    public void removeFromTree(Topic topic, String clientID) {
+        Action res;
+        do {
+            res = remove(clientID, topic, this.root, NO_PARENT);
+        } while (res == Action.REPEAT);
+    }
+
+    private Action remove(String clientId, Topic topic, INode inode, INode iParent) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.